pkg/storaged/btrfs/volume.jsx

```diff
@@ -0,0 +1,208 @@
+                                                action={
+                                                    <StorageLink onClick={rename_dialog}
+                                                               excuse={is_mounted ? _("Btrfs volume is mounted") : null}>
+                                                        {_("edit")}
+                                                    </StorageLink>}
+                    />
+                    <StorageDescription title={_("UUID")} value={uuid} />
+                    <StorageDescription title={_("Capacity")} value={fmt_size_long(use[1])} />
+                    <StorageDescription title={_("Usage")}>
+                        <StorageUsageBar key="s" stats={use} />
+                    </StorageDescription>
+                </DescriptionList>
+            </CardBody>
+            <CardHeader><strong>{_("btrfs devices")}</strong></CardHeader>
+            <CardBody className="contains-list">
+                <PageTable emptyCaption={_("No devices found")}
+                                   aria-label={_("btrfs device")}
+                                   crossrefs={get_crossrefs(uuid)} />
+            </CardBody>
+        </StorageCard>
+    );
+};
+
+const BtrfsSubVolumesCard = ({ card, volume }) => {
+    return (
+        <StorageCard card={card}>
+            <CardBody className="contains-list">
+                <ChildrenTable emptyCaption={_("No subvolumes")}
+                               aria-label={_("btrfs subvolumes")}
+                               page={card.page} />
+            </CardBody>
+        </StorageCard>
+    );
+};
+
+function make_btrfs_subvolume_pages(parent, volume) {
+    const subvols = client.uuids_btrfs_subvols[volume.data.uuid];
+    if (subvols) {
+        for (const subvol of subvols) {
+            make_btrfs_subvolume_page(parent, volume, subvol);
+        }
+    } else {
+        const blocks = client.blocks[volume.path];
+        /*
+         * Try to show subvolumes based on fstab entries, this is a bit tricky
+         * as mounts where subvolid cannot be shown userfriendly.
+         */
+        let has_root = false;
+        for (const config of blocks.Configuration) {
+            if (config[0] == "fstab") {
+                const fstab_subvol = {};
+                const opts = decode_filename(config[1].opts.v).split(",");
+                for (const opt of opts) {
+                    if (opt.startsWith("subvol=")) {
+                        fstab_subvol.pathname = opt.replace("subvol=", "");
+                    } else if (opt.startsWith("subvolid=")) {
+                        fstab_subvol.id = opt.replace("subvolid=", "");
+                    }
+
+                    if (fstab_subvol.id && fstab_subvol.pathname) {
+                        break;
+                    }
+                }
+
+                if (fstab_subvol && fstab_subvol.pathname === "/") {
+                    has_root = true;
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231208-171735-1465bfd6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)